azure/services/privatedns/privatedns.go

```diff
@@ -20,202 +20,250 @@ import (
-				"Please see https://capz.sigs.k8s.io/topics/custom-dns.html#manage-dns-via-capz-tool", "private DNS", zoneSpec.ZoneName)
-			return nil
-		}
-		// Create the private DNS zone.
-		log.V(2).Info("creating private DNS zone", "private dns zone", zoneSpec.ZoneName)
-		pDNS := privatedns.PrivateZone{
-			Location: to.StringPtr(azure.Global),
-			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-				ClusterName: s.Scope.ClusterName(),
-				Lifecycle:   infrav1.ResourceLifecycleOwned,
-				Additional:  s.Scope.AdditionalTags(),
-			})),
-		}
-		err = s.client.CreateOrUpdateZone(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, pDNS)
-		if err != nil {
-			return errors.Wrapf(err, "failed to create private DNS zone %s", zoneSpec.ZoneName)
-		}
-		log.V(2).Info("successfully created private DNS zone", "private dns zone", zoneSpec.ZoneName)
-		for _, linkSpec := range zoneSpec.Links {
-			// If the virtual network link is not managed by capz, skip its reconciliation
-			isVnetLinkManaged, err := s.isVnetLinkManaged(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, linkSpec.LinkName)
-			if err != nil && !azure.ResourceNotFound(err) {
-				return errors.Wrapf(err, "could not get vnet link state of %s in resource group %s", zoneSpec.ZoneName, s.Scope.ResourceGroup())
-			}
-			// If resource is not found, it means it should be created and hence setting isVnetLinkManaged to true
-			// will allow the reconciliation to continue
-			if err != nil && azure.ResourceNotFound(err) {
-				isVnetLinkManaged = true
-			}
-			if !isVnetLinkManaged {
-				log.V(2).Info("Skipping vnet link reconciliation for unmanaged vnet link", "vnet link", linkSpec.LinkName, "private dns zone", zoneSpec.ZoneName)
-				continue
-			}
-			// Link each virtual network.
-			log.V(2).Info("creating a virtual network link", "virtual network", linkSpec.VNetName, "private dns zone", zoneSpec.ZoneName)
-			link := privatedns.VirtualNetworkLink{
-				VirtualNetworkLinkProperties: &privatedns.VirtualNetworkLinkProperties{
-					VirtualNetwork: &privatedns.SubResource{
-						ID: to.StringPtr(azure.VNetID(s.Scope.SubscriptionID(), linkSpec.VNetResourceGroup, linkSpec.VNetName)),
-					},
-					RegistrationEnabled: to.BoolPtr(false),
-				},
-				Location: to.StringPtr(azure.Global),
-				Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
-					ClusterName: s.Scope.ClusterName(),
-					Lifecycle:   infrav1.ResourceLifecycleOwned,
-					Additional:  s.Scope.AdditionalTags(),
-				})),
-			}
-			err = s.client.CreateOrUpdateLink(ctx, s.Scope.ResourceGroup(), zoneSpec.ZoneName, linkSpec.LinkName, link)
-			if err != nil {
-				return errors.Wrapf(err, "failed to create virtual network link %s", linkSpec.LinkName)
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	zoneSpec, links, records := s.Scope.PrivateDNSSpec()
+	if zoneSpec == nil {
+		return nil
+	}
+
+	if err := s.reconcileZone(ctx, zoneSpec); err != nil {
```

Comment:
In [this PR](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2093) I discussed w/ @CecileRobertMichon that we don't want to update the put status for an unmanaged resource. In this case I think we want to add a check earlier to return or skip this block if the zone isn't managed.